### DISCUSSION TITLE
v5.53.1 increases tests time execution by ~25%

### DISCUSSION THREAD

**Discussion by janowskipio** [TanStack/query#8026] [General] [answered]
Hello again!
We've recently tried to update to the newest version of the tanstack query and during this, we've found out that the change introduced in the [v5.53.1](https://github.com/TanStack/query/releases/tag/v5.53.1), https://github.com/TanStack/query/pull/7966/files, increases our tests time execution by ~25%.

For now, we've disabled `structuralSharing` in the tanstack query config used for the test env 

```
if (process.env.NODE_ENV === 'test') {
  queryClient.setDefaultOptions({
    queries: { retry: false, structuralSharing: false },
  })
}
```

As I've not found any similar issue reported yet, I'm not sure if this issue affects everyone in the same manner, thus I've decided to start a discussion here (it may have to do with our large codebase and specific implementations).

Do you think it would be beneficial to mention the potential perf downside of using `structuralSharing` in the docs? 

Or include it in the [testing part of the docs ](https://tanstack.com/query/latest/docs/framework/react/guides/testing), to maybe even disable it by default for the tests?

Also, I see a big performance downgrade in some parts of the app itself as well - based on the specific tests that have now big execution time. I think it may be unexpected behavior for the developers - wasting time wondering what may cause such a perf problem during development mode - but I'd need to dig deeper to find out why specific parts of the app are affected that much. Although it may be connected with some of [the ~"big data" we are using](https://eclass.eu/en/eclass-standard/search-content), I don't see any direct connection with the tanstack query on the first glance.

Looking forward to hearing your opinion and ideas! Thank you for your great work 🙇

**Comment by janowskipio** [comment]
ok, seems like disabling `structuralSharing` also causes other issues to popup (so it has its own value in our project 🤝 )

FYI, for now, we have just used `replaceEqualDeep`
```
if (process.env.NODE_ENV === 'test') {
  queryClient.setDefaultOptions({
    queries: { retry: false, structuralSharing: replaceEqualDeep },
  })
}
```
to skip these expensive (in our case) changes from [v5.53.1](https://github.com/TanStack/query/releases/tag/v5.53.1)

**Comment by TkDodo** [comment]
I think stringifying `prevData` is actually unnecessary, so we could remove that:

https://github.com/TanStack/query/blob/48702e4586449e9e86ce3c02f4412b261e9801ae/packages/query-core/src/utils.ts#L359

want to file a PR ?

**Reply by janowskipio** [reply]
I'm sorry, but I'm currently overwhelmingly busy with my work (did this upgrade in between the sprints). The least I could do was to share my findings, but I won't have time this month to engage more on this topic 🤝 🙇

**Reply by TkDodo** [reply, marked as answer]
we've already shipped an improvement
